backend/api_v1/dns.go

```diff
@@ -78,252 +78,434 @@ type DNSResponse struct {
+	}
+	eg := errgroup.Group{}
+	for i, recordLoop := range recordTypes {
+		// Get all items which may not be thread safe.
+		record := recordLoop
+		packetType := recordTypesPacket[i]
+
+		// Do the DNS lookup.
+		eg.Go(func() error {
+			// Do the DNS lookup.
+			msg, err := godnsLookup(log, addr, packetType, strings.Join(chunks, ".")+".")
+			if err != nil {
+				return err
 			}
-		}
 
-		// Go through each record to make the message.
-		results := make([]*godns.Msg, len(recordTypesPacket))
-		anyQclass := godns.StringToClass["IN"]
-		wg := errgroup.Group{}
-		for i, v := range recordTypesPacket {
-			resultPtr := &results[i]
-			qtypeAndHostname := v
-			wg.Go(func() error {
-				// Make the DNS connection.
-				dnsServer := cachedDnsServer
-				if !params.Cache {
-					dnsServer = "1.1.1.1:53"
-				}
-				conn, err := godns.Dial("tcp", dnsServer)
-				if err != nil {
-					log.Error("failed to connect to dns server", zap.Error(err))
-					return err
-				}
+			// Make each response.
+			dnsResponses := make([]*DNSResponse, 0)
+		answerIteration:
+			for _, v := range msg.Answer {
+				// Handle the various responses.
+				var data json.RawMessage
+				originalValue := v
+				resultDnsHost := dnsServer
+			parseAnswer:
+				switch x := v.(type) {
+				case *godns.CNAME:
+					if record == "CNAME" {
+						// This is to be expected here since we are looking for CNAME records.
+						b, _ := json.Marshal(x.Target)
+						data = b
+					} else {
+						// In this situation, the DNS configuration is telling us to look elsewhere.
+						recursionCount := 0
+						for recursionCount < 50 {
+							// Chunkify the CNAME.
+							chunkifyReady := x.Target
+							if strings.HasSuffix(chunkifyReady, ".") {
+								chunkifyReady = strings.TrimRight(chunkifyReady, ".")
+							}
+							cnameChunks := strings.Split(chunkifyReady, ".")
 
-				// Defer killing the connection to stop leaks.
-				defer conn.Close()
+							// Get the NS host.
+							nsHost, _, _, err := findNameserverHostname(log, initAddr, cnameChunks)
+							if err != nil {
+								return err
+							}
+							if nsHost == "" {
+								// Unable to find NS record.
+								log.Warn("unable to find NS record", zap.String("hostname", strings.Join(cnameChunks, ".")))
+								continue answerIteration
+							}
 
-				// Create the DNS message.
-				msg := &godns.Msg{}
-				msg.Id = godns.Id()
-				msg.RecursionDesired = true
+							// Turn that into the address.
+							rawAddr, err := net.ResolveIPAddr("ip", nsHost)
+							if err != nil {
+								return err
+							}
+							addr := rawAddr.IP.String() + ":53"
 
-				// DNS servers prefer 1 message per request. Make the question.
-				msg.Question = []godns.Question{{
-					Name:   qtypeAndHostname.hostname,
-					Qtype:  qtypeAndHostname.recordType,
-					Qclass: anyQclass,
-				}}
+							// Lookup the CNAME's value.
+							msg, err = godnsLookup(log, addr, packetType, x.Target)
```

Comment:
Is it intentional to shadow the `msg` variable set on line 263 at the beginning of the `eg.Go()` func? This feels like something we probably don't want to do, as we're within a for loop iterating over `msg.Answer`.